### PR TITLE
clojure 1.8.0.132 (new formula)

### PR DESCRIPTION
**Pull request by puredanger**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by puredanger** [comment]
Clojure (https://clojure.org) is a programming language, which until now has not shipped with any tools. This formula provides a Clojure REPL tool (`clj`) for the last stable version 1.8 and a devel version for the upcoming 1.9 release (still in alpha). I am a member of the Clojure core team.

**Comment by puredanger** [comment]
Added a new version that provides both a `clojure` script, a general runner, and `clj`, an interactive shell. (Analogous to ruby and irb.) That required releasing a new stable and devel version. Also added caveats to help on install and improved the test.

**Comment by puredanger** [comment]
Anything I can do to help get this approved? Thanks.

**Comment by habibalamin** [comment]
@puredanger

[This commit `299bcf0`](https://github.com/Homebrew/homebrew-core/commit/299bcf0) might explain a few things.

**Comment by puredanger** [comment]
@habibalamin Hi! Thanks for the comment.

Back when that commit happened, Leiningen was definitely the best choice for getting to a Clojure repl, as the core language did not provide any script for it. This new formula (provided by the Clojure core team, of which I am a member) provides a lot more functionality for dependency management and classpath construction than the old formula and is intended for exactly the use case of getting a new user up and running with a Clojure repl and taking them the next few steps down the road. Installing this tool will be the primary instruction new Clojure users see at https://clojure.org/guides/getting_started - I have not updated the page yet as I'm waiting for the formula to be available in central.

Leiningen or Boot are other widely used Clojure build tools and are great tools for moving into full project and build management. However, I think it makes the most sense for the tools provided by and supported by the core team to provide the "clojure" formula.

Note that none of these tools are in conflict - they can all be used alongside one another and I work with all of them on a daily basis. The new clj script in the formula is not yet widely used because it's new and can't be easily obtained, but that's really a chicken-and-egg problem.

If there are questions or recommended actions, I'm happy to move this forward in whatever way necessary!

**Comment by habibalamin** [comment]
@puredanger I think there's a misunderstanding; I don't have the power to help you, I'm just a contributor of formulae, like you (that's the only reason GitHub gives me the “Contributor” badge).

I only knew where to look because I remember trying to `brew install clojure` at some point in the past and getting a message to install `lein`. When I tried it a couple of hours earlier, however, it simply printed the usual error message for non-existent formula.

It so happens that for *deleted* formulae, Homebrew prints out the commit (with its message) that deleted it, so I was able to just clone the core tap and `git show` the commit. After I confirmed the existence of this commit in the `core` tap, I simply constructed the GitHub URL to share it with you.

**Comment by puredanger** [comment]
@habibalamin ah, I see! Thanks! Maybe the info will be helpful to someone regardless!

**Comment by JCount** [comment]
One issue with this formula, as far as I can tell, is that the true installation takes place outside of the scope of the formula. Essentially, it seems as if the formula is being used to move resources from the tarball into place, with the actual install happening separately, after the formula installation, when the user first runs a command.

This means that `brew uninstall clojure` won't actually work because of the artifacts that are not tracked that stray outside of the formula's "bounds". Homebrew is designed as a package management system, not as an intermediate "bootstrapping" step in a larger workflow, (with a few clear exceptions).

Homebrew needs to be able to exert control over what gets installed and where it gets installed, simply to be able to function.

Please feel free to correct me if I am wrong.

**Comment by puredanger** [comment]
Thanks for the feedback! This script is a Clojure runner and repl. Clojure, as a language on the JVM, uses jars to form the runtime classpath. The canonical  way to obtain those jars is via the Maven API and they are typically cached to your ~/.m2 directory. This action is identical to the operations performed in the existing maven, lein, and boot formula.

If you are referring to the files installed in the ~/.clojure directory, the files there are user configuration that should survive formula updates, which seems analogous to the files many other brew formula maintain in dot directories (lein and boot are similar examples, but it seems like there are many many others). The script is designed to be tolerant of those files being absent.

If there is a different recommended approach, please let me know what it is!

**Comment by stuarthalloway** [comment]
Is additional context needed beyond what Alex said above? Happy to help move this along if there are open questions.

**Comment by ilovezfs** [comment]
This is what JCount has in mind (slightly modified version of https://github.com/JCount/homebrew-test-formulae/blob/master/Formula/clojure.rb):
```
class Clojure < Formula
  desc "The Clojure programming language"
  homepage "https://clojure.org"
  url "https://github.com/clojure/clojure/archive/clojure-1.8.0.tar.gz"
  sha256 "b0d92438a2ab6c396360d70569d07520c1edd6c6d895d6536c7d19ced72c9fe8"

  depends_on :ant => :build
  depends_on "maven" => :build
  depends_on :java => "1.7+"

  def install
    system "mvn", "package"

    libexec.install Dir[buildpath/"target/clojure-*"]

    (bin/"clj").write <<-EOS.undent
      #!/bin/bash
      exec java -cp #{libexec}/clojure-#{version}.jar clojure.main "$@"
    EOS

    bin.install_symlink "clj" => "clojure"
  end

  test do
    %w[clj clojure].each do |clj|
      assert_equal "2", shell_output("#{bin}/#{clj} -e \"(+ 1 1)\"").strip
    end
  end
end
```

**Comment by puredanger** [comment]
This is far less useful than what I have submitted, which adds a prior step that constructs the classpath with which to launch clojure.main (based on your user and project level dependency declarations). In the next Clojure release, this is actually essential because Clojure itself has jar dependencies which must be downloaded and added to the classpath. 

To reiterate that, the proposed script WILL NOT work with the next version of Clojure. It is essential to have some way to build a classpath based on jar dependencies and that is exactly what the script in the proposed formula does. For more info, there is a forthcoming guide about the script at https://github.com/clojure/brew-install/blob/master/docs/guide-clj.adoc (will be moved into the main Clojure web site).

**Comment by ilovezfs** [comment]
We will have to pass on this formula unless it is able to install everything it needs in the Cellar, not just install an install script that is invoked after the fact.

**Comment by ilovezfs** [comment]
Basically, the stuff that your `install-clj` is fetching into the user's ~/.m2 should instead be fetched during `def install` into `libexec`, and then a wrapper script should point `clj` to that location.

**Comment by puredanger** [comment]
I could integrate the install-clj script into the clojure script if that would be better - it's mostly broken out for convenience of implementation. Maybe I have just named that in a misleading way? All the `install-clj` script is doing is initializing the ~/.clojure directory if it is not available. The final step in the script is just ensuring that the key libraries have been downloaded and cached in your Maven repository.

This seems analogous to many other existing formula which install config into a dot directory. I don't really see any difference between what is happening here and what most other brew formula do?

I would really like to make this work as we have thousands of Clojure users that will really benefit from having this readily available in the central tap.

**Comment by ilovezfs** [comment]
Let's say a user runs `brew upgrade clojure` to upgrade the formula. How do the "the key libraries [that] have been downloaded and cached in your Maven repository" then get upgraded, and when does that happen?

**Comment by habibalamin** [comment]
I really appreciate that the Homebrew team are very careful to make sure formulae aren't just installing install scripts and other annoying antipatterns, but I don't see how your question is relevant, @ilovezfs. I think what you're getting at — correct me if I'm wrong — is that if the libraries aren't updated, then the formula upgrade wasn't really an upgrade.

I guess this depends; are the libraries basically the Clojure standard libraries, and if so, is the Clojure version tied to the standard library version? If so, you could get into a situation where you upgrade the Clojure formula so that Homebrew thinks it's on version X.2, but the libraries in the Maven path aren't updated, which means Clojure is actually on version X.1 (because it's using an older version of the standard library).

I'm not a Clojure user, for the record, but I am a Homebrew user, and package managers are, almost by definition, supposed to have 99% of their “features” not used by 99% of their users (if you consider each package a feature (of collectively diminishing returns), which, given that package selection is a quite important metric of the usefulness of a package manager, I would (in this context; it may not be technically accurate or exactly what I mean, but I think it's a useful mental model)).

**Comment by ilovezfs** [comment]
>I think what you're getting at — correct me if I'm wrong — is that if the libraries aren't updated, then the formula upgrade wasn't really an upgrade.

Right.

**Comment by puredanger** [comment]
Sorry for the delay in response, we've been discussing! The "install-clj" script and the Java code it invokes is doing nothing more than determining the libraries needed to run the classpath construction part of the clj script and ensure they exist on the system. It's currently doing so by invoking the Maven libraries (aka Aether) to ensure the necessary libraries are downloaded and cached in the user's Maven repository (which is the typical way that JVM libraries are managed). 

However, we could instead include all of the libraries needed by the clj script in the tar file and install them into the Cellar where brew can manage it for upgrade or uninstall. Would this be ok?

We still need a user config directory that should survive across brew's upgrade cycle. I would prefer to keep the construction of that in the clj script (if needed), but could also create it explicitly during the brew install if that would be preferred. I have tried to find other formula doing things like this but have not found any examples to draw from. Would it be ok to leave that in the clj script?

I'd be happy to move forward with the changes, but I'd like to know whether this direction seems like it will address the problems.

Thanks for working through this with this, I really appreciate it.

**Comment by ilovezfs** [comment]
>However, we could instead include all of the libraries needed by the clj script in the tar file and install them into the Cellar where brew can manage it for upgrade or uninstall. The install-clj script would then go away. Would this be ok?

Yes that would be good.

>We still need a user config directory that should survive across brew's upgrade cycle. I would prefer to keep the construction of that in the clj script (if needed), but could also create it explicitly during the brew install if that would be preferred. I have tried to find other formula doing things like this but have not found any examples to draw from. Would it be ok to leave that in the clj script?

Typically we put config files in `etc` (or `etc/"clojure"`), which is actually `/usr/local/etc`, not `#{prefix}/etc` (i.e., `etc` is _not_ in the Cellar), and it persists across upgrades. `etc.install` will try to do the right thing and preserves existing config files rather than overwriting. There's a special exemption in the sandbox that allows you to tweak them in `def post_install` if needed.

That said, some things do use `~/.foo` config directories that are created on demand and that's OK too, but `etc` is preferred when possible.

**Comment by puredanger** [comment]
Thanks. For the config file, this is user-specific, so it seems like putting it in etc would not be the right level of sharing. In that case, somewhere under $HOME seems better.

**Comment by ilovezfs** [comment]
That's fine :)

**Comment by JCount** [comment]
@puredanger Feel free to use `maven` directly as a dependency if it would make constructing the formula's `install` block any easier. (If that would decrease complexity of the tarball you are planning on using, etc. However, it might be easier for upstream (you) to directly manage that.) :)

If you have any questions, please ask.

**Comment by puredanger** [comment]
I have considered using Maven as a dependency. However, many Clojure users do not use Maven directly (only indirectly via programs that use its libraries) and I hate to levy that dependency on them if they don't require it. 

My plan is to build a single uber jar containing all necessary dependencies and install that in the cellar, so the tarball and formula itself should be easy.

**Comment by JCount** [comment]
Well, then, sounds like you have a plan. :smile: 

To clarify my earlier statement, which was essentially a poorly worded version of what @ilovezfs later conveyed. Namely that in its current state it would integrate well with the Homebrew ecosystem; namely the standard upgrade, reinstall, and uninstall commands wouldn't necessarily function as expected.

_That_ is the essence of what I meant.

**Comment by puredanger** [comment]
Thanks, working on it now!

**Comment by ilovezfs** [comment]
I'd like to see the steps in the install method wrapped up into something that can either do

```
./configure --prefix=#{prefix}
make install
```
or
```
make install PREFIX=#{prefix}
```
or
```
./install.sh --prefix=#{prefix}
```

**Comment by ilovezfs** [comment]
(preferably `./install.sh --prefix=#{prefix}` so that we don't need to have bottles)

**Comment by puredanger** [comment]
Do you mean these steps? 

```ruby
    prefix.install Dir["*.jar"]
    prefix.install "deps.edn"
    inreplace "clojure", /PREFIX/, prefix
```

I thought the bin install also takes care of linking, etc?

**Comment by ilovezfs** [comment]
We want the logic for installation to live in a script you control upstream not in the formula so that e.g. if another executable is added the formula's install method won't need to be changed.

**Comment by puredanger** [comment]
how do I know where bin is from the script? should I just assume it's under prefix?

**Comment by ilovezfs** [comment]
Also note the payload of jars should go in libexec not prefix

**Comment by ilovezfs** [comment]
>how do I know where bin is from the script? should I just assume it's under prefix?

Yes. The standard install locations are described here: https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md#variables-for-directory-locations

So basically
```
/usr/local/Cellar/clojure/1.8.0.119/bin
/usr/local/Cellar/clojure/1.8.0.119/libexec
/usr/local/Cellar/clojure/1.8.0.119/share/man
```
and `prefix` is passed to your script as `/usr/local/Cellar/clojure/1.8.0.119`

**Comment by puredanger** [comment]
@ilovezfs how's that?

File: Formula/clojure.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    version "1.9.0-alpha19.166"
+  end
+
+  bottle :unneeded
+
+  depends_on :java => "1.7+"
+  depends_on "rlwrap"
+
+  def install
+    system("./install.sh #{prefix}")
```
Comment (JCount) [inline Formula/clojure.rb]: @puredanger this can simply be:
```
system "./install.sh", "#{prefix}"
```

**Comment by JCount** [comment]
I apologize, I feel like I am leading you in circles. `"#{prefix}"` can simply be `prefix`, no string interpolation needed.

**Comment by puredanger** [comment]
Just checking in - anything else I need to do?

**Comment by puredanger** [comment]
Good morning, anything I can do here?

**Comment by ilovezfs** [comment]
Thanks for your first contribution to Homebrew, @puredanger! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by puredanger** [comment]
Thanks for your help and time!

**Pull request merged**